docs/languages/csharp-iterators.md

```diff
@@ -0,0 +1,247 @@
+}
+```
+ 
+Look at the methods above. The first uses the standard `return` statement to return
+either an empy collection, or the iterator created by the second method. The second
+method uses the `yield return` statement to create the requested sequence.
+
+# Deeper Dive into `foreach`
+
+The `foreach` statement expands into a standard idiom that uses the
+`IEnumable<T>` and `IEnumerator<T>` interfaces to iterate across all
+elements of a colleciton. It also  minimizes errors developers make
+by not properly managing resources. 
+
+The compiler translates the `foreach` loop shown in the first
+example into something similar to this construct:
+
+```cs
+IEnumerator<int> enumerator = collection.GetEnumerator();
+while (enumerator.MoveNext())
+{
+    var item = enumerator.Current;
+    Console.WriteLine(item.ToString());
+}
+```
+
+The construct above represents the code generated by the C# compiler as of
+version 5 and above. Prior to version 5, the `item` variable had a different scope:
+
+```cs
+// C# versions 1 through 4:
+IEnumerator<int> enumerator = collection.GetEnumerator();
+int item = default(int);
+while (enumerator.MoveNext())
+{
+    item = enumerator.Current;
+    Console.WriteLine(item.ToString());
+}
+```
+
+This was changed because the earlier behavior could lead to subtle and hard
+to diagnose bugs involving lambda expressions. See the section on
+[lambda expressions](index.md) for more information. note: redirecting to the
+index seems the best way to handle a reference to a doc not yet written. I'm open
+to other suggestions.
+
+The exact code generated by the compiler is somewhat more complicated, and
+handles situations where the object returned by `GetEnumerator()` implements
+the `IDisposable` interface. The full expansion generates code more like this:
+
+```cs
+{
+    var enumerator = collection.GetEnumerator();
+    try {
+        while (enumerator.MoveNext())
+        {
+            var item = enumerator.Current;
+            Console.WriteLine(item.ToString());
+        }
+    } finally {
+        // dispose of enumerator.
+    }
+}
+```
+
+The manner in which the enumerator is disposed of depends on the characteristics of
+the type of `enumerator`. In the general case, the `finally` clause expands to:
+
+```cs
+finally {
+   (enumerator as IDisposable)?.Dispose();
+} 
+```
+
+However, if the type of `enumerator` is a sealed type and there is no implicit
+conversion from the type of `enumerator` to `IDisposable`, the `finally` clause
+expands to an empty block:
+```cs
+finally {
+} 
+```
+
+If there is an implicit conversion from the type of `enumerator` to `IDisposable`,
+and `enumerator` is a non-nullable value type, the `finally` clause expands to:
+
+```cs
+finally {
+   ((IDisposable)enumerator).Dispose();
+} 
+```
+
+Thankfully, you don't need to remember all these details. The `foreach` statement
+handles all those nuances for you. The compiler will generate the correct code for
+any of these constructs. 
+
+
```

Comment:
I'm thinking we should link to a full sample (that we create in the samples folder) and probably could pull in the sample as the reference code in this tutorial. Eventually, we want to get to the point where the samples are tested against the core builds. I'm not exactly sure what the md syntax is for pulling in lines of code though. /cc @cartermp